### PR TITLE
Overload HbaseIO with KV<RowKey, RowMutations>

### LINKED ISSUE #25830: [Feature Request]: Overload HbaseIO with KV<RowKey, RowMutations>
### What would you like to happen?

We want to add HbaseIO support for writing KV<RowKey, RowMutations> to Hbase. We anticipate using this in Dataflow for writing Bigtable change streams to Hbase. Support for the KV signature will allow atomic, in-order mutations to be applied to Hbase *most of the time*. 

### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [X] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by georgecma**
Add HbaseIO support for writing RowMutations (ordered by rowkey) to Hbase.
 Fixes #25830 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
test not run due to spotbug: HbaseRowMutationIOTest.java:56 "You should name sfl4j Loggers as LOG."

File: sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java [hunk 1]
```diff
@@ -765,4 +796,195 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+                      connection.getAdmin().isTableAvailable(TableName.valueOf(tableId))),
+                  Boolean.toString(
+                      connection.getAdmin().isTableEnabled(TableName.valueOf(tableId))),
+                  "\nConnection Closed/Aborted/Locks:",
+                  Boolean.toString(connection.isClosed()),
+                  Boolean.toString(connection.isAborted()))));
+        }
+
+        // Dummy output so that we can get Dataflow stats for throughput.
+        c.output(1);
```
Comment (Abacn) [inline sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:977]: dummy output may not be suitable in production code. Either implement WriteResult or use PDone for now (and implement WriteResult as future task)

File: sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java [hunk 2]
```diff
@@ -765,4 +796,195 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        if (table != null) {
+          table.close();
+          table = null;
+        }
+
+        LOG.debug("Wrote {} records", recordsWritten);
```
Comment (Abacn) [inline sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:936]: This is always 0, looks like missing a line of incrementing recordsWritten somewhere?

**Review by Abacn — COMMENTED** [review]
Thanks, looked at HBaseIO but not yet tests and utility classes. 

By saying "in-order write", what is the requirement? Is that something like <Key=1, RowMutations1> will be written prior to <Key=2, RowMutations1> ? Or just Mutations in a `RowMutations` ("Performs multiple mutations atomically on a single row.") should be executed in order?

**Comment by aromanenko-dev** [comment]
@georgecma Thanks for contribution! I see that this PR is created from your local `master` branch. Please, create and use a feature branch for this.

File: sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java (hunk 2 above)
Comment (georgecma) [reply sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:936]: Added in processElement

File: sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java (hunk 1 above)
Comment (georgecma) [reply sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:977]: Using PDone for now

File: sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java [hunk 3]
```diff
@@ -150,6 +155,32 @@
...
+ * Configuration configuration = ...;
+ * PCollection<KV<byte[], RowMutations>> data = ...;
+ *
+ * data.apply("write",
+ *     HBaseIO.writeRowMutations()
+ *         .withConfiguration(configuration)
+ *         .withTableId("table"));
+ * }</pre>
+ *
+ * <p>Note that the transformation emits the number of RowMutations written as an integer after
```
[the remark this answers is not in the document]
Comment (georgecma) [reply sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:181]: Using PDone for now to be consistent with rest of HBaseIO

**Comment by georgecma** [comment]
> Add HbaseIO support for writing RowMutations (ordered by rowkey) to Hbase.
> Fixes #25830


Thanks, created new branch after responding to @Abacn's comments. New PR is https://github.com/apache/beam/pull/25864 
Closing this PR

**Pull request closed without merging**